### ISSUE TITLE
Umbrella: post-U8 follow-up design queue before autonomous implementation

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#1489]
This is the coordination issue for the post-U8 follow-up design queue after PR #1488, starting from `main` at:

```text
4417100ffbd7c1098cab50ddf262534d0e9a4e42
```

Review decision: do not run this as one large autonomous implementation PR. Split the queue, and do not implement static-audit performance candidates until a profile or representative end-to-end benchmark proves they are worth the semantic, aliasing, cache, or maintenance risk.

Benchmark campaign tracker: #1490.

## Accepted PR partition

0. **PR 0: #1491 repository-rules routing cleanup**
   - Restructure `REPOSITORY_RULES.md` as review-bot routing configuration before the implementation queue starts.
   - Fix the stale extension registration rule so the bot no longer teaches retired `register_runtime` guidance.
   - Split the oversized `Performance And Layout Rules` section into routed sections that match `scripts/repository-rules-review.py` triggers.
   - Add the accepted #1489 profile-first and aliasing/uniqueness policies once, in their correct routed/process sections.
   - Add consistency coverage so every `##` section is routed or explicitly marked human-only.
   - This is a rules/tooling PR, not a product implementation PR.

1. **PR 1: #1487 and #1468**
   - #1487: move TBLIS source-build support out of `tenferro-cpu` into an external `dot_general` provider example.
   - #1468: remove remaining linalg semantic AD legacy rule-recording reuse where equivalent semantic linearize/transpose emission is available.
   - This PR is packaging plus correctness cleanup, not runtime architecture or performance tuning.

2. **PR 2: #1456 and #1471 residuals**
   - Do not redo the U2/U3 pieces already covered by #1488.
   - Remaining U2: narrow the public `PreparedOperation::execute` hook into an internal/sealed execution boundary or accepted equivalent, keep prepared-operation public metadata small, preserve `run_compiled_values` lazy-view behavior, and document the owned/view/current-slot value representation.
   - Remaining U3: explicit scheduled transfer nodes, split-use buffer lifetime semantics, native event/stream/queue mapping for CUDA and wgpu/CubeCL, handle drop semantics beyond the current join-handle contract, and fuller admission/lock-order coverage.

3. **PR 3 and later: performance batches only after gates pass**
   - #1485 and #1484 may proceed without the share gate because repeated same-shape cuTENSOR/FFT plan construction is a known fixed cost, but they still require before/after benchmarks and non-regression checks.
   - #1479, #1480, #1481, and #1482 require the profile-first entry gate below.
   - #1483 is closed as not planned unless a future profile shows gradient accumulation at `>=5%` of backward wall time.
   - #1486 stays last and remains measure-first for release-codegen splitting.

## Profile-first rule

Static/source audit findings are not sufficient justification for performance implementation. For #1479, #1480, #1481, and #1482:

- first measure the targeted path's share in a predeclared U0 workload or representative tensor-network workload;
- require the targeted path to occupy at least `5%` of wall time before implementation starts;
- if the share is below threshold, record the measurement or argument in the issue and close or defer without implementation;
- use microbenchmarks only after need is established, because a microbenchmark shows effect, not need.

This rule should land through #1491 in `REPOSITORY_RULES.md` under the correct process/routing section so future static audits do not resurrect unproven risky optimizations.

## Queue benchmark policy

#1490 pins the queue-start baseline and carries the #1462 watchlist forward.

For this cleanup stream:

- block on correctness/oracle failures;
- block on a valid paired benchmark whose candidate-relative 95% upper bound exceeds `+20%` on a predeclared primary workload;
- record `+5%` to `+20%` changes as watchlist items unless the final campaign shows a reproducible attributed regression;
- do not relax thresholds after seeing results.

## Cross-cutting aliasing rule

Add this rule through #1491 to `REPOSITORY_RULES.md` before implementing alias-increasing or in-place-write optimizations:

> An in-place write into a buffer reachable through an `Arc` is allowed only under (a) a uniqueness proof at the write site (`Arc::get_mut` / `Arc::try_unwrap` / `Arc::make_mut`, or an `Arc::ptr_eq` alias rejection for backend buffers), or (b) exclusive runtime ownership of the buffer for its whole lifetime, as in dispatch-loop slots.

Consequences:

- #1480: every internal handoff converted from copy to view needs mutation-after-handoff coverage.
- #1479 and #1480 must not be reviewed in isolation when aliasing and strided in-place accumulation interact.
- #1483 remains closed unless a profile justifies reopening; if reopened, accumulation must be copy-on-write and must preserve user-held `grad()` handles.

## Issue-specific constraints

- #1487: keep `tenferro-cpu` as the public CPU crate. TBLIS is an external `dot_general` provider only. Decide whether existing `cpu-tblis` / `cpu-tblis-runtime` / `cpu-tblis-linked` features are removed immediately or kept through a deprecation window, and ensure the moved provider is built in CI if it remains in-tree.
- #1468: close the “where practical” escape hatch. No `RecordedBuilder` family imports should remain under `crates/tenferro-linalg/src/ad/`, or each remaining site must carry a concrete recorded justification. `ad_support_manifest` and oracle replay classifications must remain unchanged unless support genuinely changes.
- #1481: a fused multi-axis reduction may change floating-point accumulation order. Decide against determinism policy and oracle tolerances before implementation.
- #1485: do not key cached cuTENSOR plans on allocation-specific pointer alignment. Key required alignment, verify actual buffers at execution, and replan or reject on mismatch. Device workspace retained bytes must be accounted separately, with bounded eviction behavior.

## Inplace indexing design decision

The current `docs/design/inplace-indexing.md` should be updated before implementation starts from it. Accepted direction: follow the JAX-style semantic model, not PyTorch reference semantics.

- `x[idx] = value` records as a pure `index_put(x, idx, value)` operation in semantic IR.
- Execution may mutate in place only when runtime/eager ownership proves uniqueness.
- No user-visible shared mutable host storage or load-bearing version-counter semantics should be introduced.
- AD remains the ordinary scatter/gather adjoint of a pure operation.

## Completion criteria

This umbrella can close when #1491 has installed the durable routing/rules cleanup, the accepted split is reflected in child issues and implementation PRs, completed child issues are closed by their PRs, and deferred items are explicitly closed, kept open with gates, or refiled with narrower scope.

**Comment by shinaoka** [comment]
## Design review of the post-U8 follow-up queue

Reviewed against `main` at `4417100f`. Premise check first: the residual scoping in this umbrella is accurate. I verified in code that #1488 implemented the U2 core (`execute_scheduled_slots` walks `ScheduledNode`s with per-operation executors and storage tracking, cross-storage placements fail typed, `ScheduledNode::Transfer` is rejected with "not implemented before U3"), that `Runtime::submit` is currently a thread-spawn join-handle wrapper, and that the public `PreparedOperation::execute` hook with a default `Err` sits on the public trait at `runtime/capability.rs:261`. The stated U2/U3 residuals match reality, and narrowing the execute hook into a sealed boundary is the right call.

The queue needs changes in three areas before autonomous implementation starts: process shape, performance policy, and aliasing safety. Two policy decisions were made during this review and are recorded here as accepted direction.

### 1. Process: take completion option (b) and split the queue

Completion option (a) reads as one autonomous implementation PR closing the included child issues. This queue spans four kinds of work: packaging (#1487), AD correctness (#1468), runtime architecture (#1456/#1471 residuals), and seven performance items. One PR covering all of it recreates the reviewability failure this project already paid for once: the #1453 restaging regression passed review invisibly inside a 70k line diff, and that lesson is written into the U0 protocol. Take option (b) from the start:

- PR 1: #1487 and #1468 (independent, low risk)
- PR 2: #1456 and #1471 residuals (architecture, including the sealed execute boundary and the value representation documentation)
- PR 3 and later: performance items in batches of two or three, each carrying its own benchmark evidence

PR #1488 (+3.8k) is a good reference for the upper size bound.

### 2. Accepted policy: performance work requires numerical justification before implementation

Decision: optimizations not backed by measurements are not implemented.

The performance items #1479 through #1485 descend from the #1426 static audit. Their "Current evidence" sections cite code structure, not profiles. Per-item before/after Criterion evidence does not fix this, because a microbenchmark of the touched path always improves: that is evidence of effect, not evidence of need. Need means the candidate path's share of an end-to-end workload.

Note the asymmetry in the current queue: #1486 is the only item that imposes "measure first" on itself, and it is about build time. The same discipline must apply to the runtime items. Concretely:

- Entry gate for each of #1479, #1480, #1481, #1482, #1483: first measure the candidate path's share in a predeclared U0 workload or a tensor-network representative benchmark. Below an agreed share threshold (proposal: 5 percent), record the measurement in the issue and close it without implementation.
- #1484 and #1485 are exempt from the share gate: plan construction cost for FFT and cuTENSOR is an externally known, large fixed cost, and repeated same-shape contraction is the central tensor-network pattern. They still need before/after evidence, but their motivation is already established.
- #1483 has the weakest case in the queue. Eager backward was the most improved region in the U8 measurements (shape churn around minus 50 percent), gradient accumulation already goes through the buffer pool, and no measurement shows accumulation allocation as a bottleneck. It also carries the highest semantic risk (section 4). It should not be implemented ahead of a profile that justifies it.

Resulting order: #1487, #1468, #1456/#1471 residuals, then #1485 and #1484, then profile, then whichever kernel items (#1479/#1481/#1482) survive the share gate, and #1480/#1483 only if both the profile justifies them and the aliasing rule below is in place. #1486 stays last, as already written.

### 3. Benchmark protocol: define the threshold and create the campaign tracker

The constraint "block on clear or extreme regressions" has no number. #1462 closed using a +20 percent upper bound, which was a post-hoc relaxation of the predeclared U0 thresholds (+5/+10 percent). Before the queue starts:

- Predeclare the block threshold for this queue.
- Create a tracking issue for the final formal benchmark campaign referenced by the constraints. It currently has no tracker: #1462 is closed and its watchlist (the broadcast elementwise cases) lives only in comments on a closed issue, which is how the #1457 constraints got lost last time. Move the watchlist into the new issue body.
- Pin the queue-start baseline at `4417100f`. Per-item before/after numbers do not compose across seven items; the cumulative effect must be measured at the end against the queue-start pin.

### 4. Cross-cutting aliasing rule (missing from the constraints)

The cross-cutting constraints cover cache ownership, column-major semantics, and thread policy, but say nothing about aliasing, while three queued designs erode the property that currently keeps the codebase safe. Today safety comes from two structural facts: host buffers are `Buffer::Host(Vec<T>)` with unique ownership, so host sharing exists only at `Arc<Tensor>` granularity, and there are essentially no user-facing in-place operations. That is safety by absence, not by mechanism. #1480 increases the number of aliases (identity handoffs preserving views), #1479 makes in-place accumulation into strided outputs faster and more common, and #1483 introduces in-place writes into slots whose handles escape through `grad()` (which returns `Option<Arc<Tensor>>`, a shared handle, `eager.rs:2993`).

One more structural fact: `Buffer::Backend(Arc<dyn BackendBuffer<T>>)` means `&mut Tensor` does not imply a unique device allocation. The host intuition that `&mut` equals exclusivity is false on GPU. The CubeCL `copy_into` path already defends against this manually with an `Arc::ptr_eq` alias rejection (`cubecl/mod.rs:961`), and the slot reclaim path already gates pool reuse on `Arc::try_unwrap` (`runtime/src/exec.rs:1053`). Those two sites are the correct pattern; the rule below generalizes them.

Proposed addition to the constraints (and to REPOSITORY_RULES so future agents inherit it):

> An in-place write into a buffer reachable through an `Arc` is allowed only under (a) a uniqueness proof at the write site (`Arc::get_mut` / `Arc::try_unwrap` / `Arc::make_mut`, or an `Arc::ptr_eq` alias rejection for backend buffers), or (b) exclusive runtime ownership of the buffer for its whole lifetime, as in the dispatch-loop slots.

Additionally, a debug-build-only mutation assertion (a version counter that is compiled out in release) on `TypedTensor` would make violations of this rule mechanically detectable in tests. This is an assertion, not load-bearing semantics; see section 6 for why it must never become the latter.

Issue-specific consequences:

- **#1483**: accumulation must be copy-on-write. `Arc::make_mut` is the exact idiom: in-place when the slot is the unique owner, silent clone when a user holds a `grad()` handle. Uniqueness must be checked at the storage level, not only on the outer `Arc<Tensor>`. Acceptance criterion to add: hold a `grad()` handle, run `backward()` again, assert the held tensor is unchanged.
- **#1480**: every internal handoff converted from copy to view needs a mutation-after-handoff test (mutate downstream, assert the upstream holder is unchanged). Note the interaction with #1479: one issue adds aliases while the other adds in-place writes; they must not be reviewed in isolation.

### 5. Issue-specific design gaps

- **#1485**: pointer alignment cannot be part of a static cache key, because alignment varies per allocation. Key the required alignment, verify the actual buffers against it at execution time, and fall back to replanning on mismatch. Cached workspaces hold device memory: retained-byte accounting must report device bytes separately, and an eviction policy under memory pressure is required even with bounded defaults. Context and handle lifetime is already covered in the issue.
- **#1481**: changing multi-axis reduction from axis-at-a-time to a fused traversal changes floating-point accumulation order and therefore bit-level results. Decide against the design document's determinism policy and the oracle replay tolerances whether order changes are permitted, before implementation. The same applies to any accumulation-order change in #1483.
- **#1468**: "where practical" is an escape hatch. Close it with a behavioral bar: no `RecordedBuilder` family imports remain under `crates/tenferro-linalg/src/ad/`, or each remaining site carries a recorded justification, and `ad_support_manifest` plus oracle replay classifications are unchanged.
- **#1487**: the direction (external `dot_general` provider, CPU backend stays) is right. Two open decisions: what happens to the existing `cpu-tblis` / `cpu-tblis-runtime` / `cpu-tblis-linked` features (immediate removal or deprecation cycle), and whether `ext/tenferro-cpu-tblis` is built in CI. If ext examples are outside CI, the moved provider becomes unverified code the moment it moves.

### 6. Accepted direction: inplace indexing follows JAX, not PyTorch

This is outside the queue but shares its root with the #1480/#1483 aliasing question, and it should be settled before anyone implements from the current design document.

`docs/design/inplace-indexing.md` takes PyTorch (`index_put_`, view plus `copy_`, version counters, view/in-place autograd checks) as the reference model. That model is the consequence of PyTorch's reference semantics, where views share storage and in-place writes are visible through them; the version counter exists to detect saved-for-backward residuals invalidated by such writes. Adopting it in tenferro would require introducing shared mutable host storage that the current design deliberately does not have.

Decision: align with JAX instead. `x[idx] = value` records as a pure `index_put(x, idx, value)` operation in the semantic IR, and executes in place only when the runtime proves buffer uniqueness (last-use donation on the traced path, reference-count check on the eager path). JAX's existence demonstrates that no version counter is needed when mutation is expressed functionally. The building blocks already exist in this codebase: last-use analysis with `Arc::try_unwrap` gating in the slot reclaim path, and the uniqueness idioms from section 4. The AD story reduces to an ordinary scatter/gather adjoint for one pure operation, with no view/in-place special cases. The design document should be updated to this reference model before implementation. The project's early API drew on PyTorch; the semantic-program architecture it has since converged on is the JAX-shaped half, and this is a place where the reference model should follow the architecture.

### Minor

- `runtime/schedule.rs` still opens with a module-wide `#![allow(dead_code, reason = "Phase 5 establishes schedule metadata consumed incrementally by later phases")]`. Production now consumes it, so both the allow and its reason are stale. One line in the #1456 residual.
- The ordering choices already in the umbrella are endorsed: #1487 first (packaging blocks the rest), #1486 last (measure first), runtime residuals before performance tuning (do not tune a path that is about to move).

### Summary of required changes to this umbrella

1. Switch the completion criterion to the split (option b) with the PR partition above.
2. Add the profile-first entry gate for #1479, #1480, #1481, #1482, #1483, with the share threshold; exempt #1484/#1485.
3. Predeclare the block threshold, create the final campaign tracker with the #1462 watchlist, pin the queue baseline at `4417100f`.
4. Add the aliasing rule to the cross-cutting constraints and to REPOSITORY_RULES; add the acceptance tests to #1480 and #1483.
5. Add the issue-specific items from section 5 to #1485, #1481, #1468, #1487.
6. Record the JAX-aligned inplace-indexing decision and update `docs/design/inplace-indexing.md` before any implementation starts from it.

**Comment by shinaoka** [comment]
Update: #1483 is closed as not planned under the profile-first rule from the review above. The accumulation share of backward is asymptotically 1/k of the VJP work per operation (about 1/chi for tensor-network sweeps), and the slot-level accumulation the issue targeted fires only O(#leaves) times per backward, so it fails the share gate by argument alone. The closing comment on #1483 records the full rationale and the reopen condition (a profile showing 5 percent or more of backward wall time). The performance queue for this umbrella is therefore #1479, #1480, #1481, #1482 behind the share gate, plus the exempt #1484 and #1485, with #1486 last.

**Comment by shinaoka** [comment]
Post-U8 follow-up PR batch is merged.

Merged PRs:

- #1492 -> 91bb1863e6b8616b178ae85b9693a36580d7509d
- #1493 -> 75b8af425c4cbb7d06077885101c5834e90ea2fe
- #1494 -> 5ff72a24bbf4ecbc60ca02005282040ef4e8aaae
- #1495 -> 5695fde03d8ecfaab6c2f07bb3f3a7153306d31a
- #1496 -> 240a8930e8be6ee9122250f8da5537b116726123
- #1497 -> d7d2ae462f4333a998b2a37f577950df9252bba6
- #1498 -> f6ad37943c80bb805cc140c8f00df23c1fd81175

Completed children closed through their merged PRs: #1491, #1487, #1468, #1485, #1484, and #1486.

Remaining open items are intentionally tracked outside this umbrella:

- #1490: final cumulative post-U8 benchmark campaign.
- #1456 and #1471: residual production dispatch / async submit / transfer / event-domain work.
- #1479, #1480, #1481, #1482: profile-gated performance candidates; no implementation until the target path reaches the recorded share gate.
- tensor4all/strided-rs#148: upstream ownership decision for prepared/fused strided CPU kernels before any further tenferro CPU kernel split.

This satisfies the umbrella completion criteria: durable rules routing landed, completed children are merged/closed, and deferred items are explicitly kept open with their gates or refiled upstream.

**Comment by shinaoka** [comment]
Closing: the umbrella PR batch is merged and remaining work is tracked by the residual/profile-gated/final-campaign issues listed above.
